### PR TITLE
Add the `DataContainer::getActiveRecord()` method

### LINKED ISSUE #7404: Selection of frontend module of type "root_page_dependent_modules" is somehow broken in the backend
### Affected version(s)

5.3.x

### Description

Steps to reproduce:
LAYOUT -> Themes -> Icon "Frontend modules" -> New -> select module type "Root page dependent modules"

Error in /var/logs:

```
[2024-08-01T22:24:59.871823+02:00] request.INFO: Matched route "contao_backend". {"route":"contao_backend","route_parameters":{"_route":"contao_backend","_scope":"backend","_controller":"Contao\\CoreBundle\\Controller\\BackendController::mainAction"},"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","method":"POST"} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"POST"}
[2024-08-01T22:24:59.933425+02:00] php.ERROR: Warning: foreach() argument must be of type array|object, null given {"exception":"[object] (ErrorException(code: 0): Warning: foreach() argument must be of type array|object, null given at /var/www/test/vendor/contao/core-bundle/contao/widgets/RootPageDependentSelect.php:31)"} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"POST"}
[2024-08-01T22:24:59.941158+02:00] contao.general.INFO: Version 1 of record "tl_module.id=5" has been created (parent records: tl_theme.id=1) {"contao":{"Contao\\CoreBundle\\Monolog\\ContaoContext":"{\"func\":\"Contao\\\\Versions::create\",\"action\":\"GENERAL\",\"username\":\"MyUserName\",\"browser\":\"Mozilla\\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/127.0.0.0 Safari\\/537.36 Edg\\/127.0.0.0\",\"uri\":\"https:\\/\\/abc.tld\\/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module\",\"pageId\":null}"}} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"POST","contao":{"Contao\\CoreBundle\\Monolog\\ContaoContext":"{\"func\":\"Contao\\\\Versions::create\",\"action\":\"GENERAL\",\"username\":\"MyUserName\",\"browser\":\"Mozilla\\/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit\\/537.36 (KHTML, like Gecko) Chrome\\/127.0.0.0 Safari\\/537.36 Edg\\/127.0.0.0\",\"uri\":\"https:\\/\\/abc.tld\\/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module\",\"pageId\":null}"}}
[2024-08-01T22:24:59.977194+02:00] messenger.INFO: Stopping worker. {"transport_names":["contao_prio_high"]} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"POST"}
[2024-08-01T22:25:00.048603+02:00] request.INFO: Matched route "contao_backend". {"route":"contao_backend","route_parameters":{"_route":"contao_backend","_scope":"backend","_controller":"Contao\\CoreBundle\\Controller\\BackendController::mainAction"},"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","method":"GET"} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"GET"}
[2024-08-01T22:25:00.113177+02:00] php.ERROR: Warning: foreach() argument must be of type array|object, null given {"exception":"[object] (ErrorException(code: 0): Warning: foreach() argument must be of type array|object, null given at /var/www/test/vendor/contao/core-bundle/contao/widgets/RootPageDependentSelect.php:31)"} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"GET"}
[2024-08-01T22:25:00.157863+02:00] messenger.INFO: Sending message Contao\CoreBundle\Messenger\Message\SearchIndexMessage with contao_prio_low sender using Symfony\Component\Messenger\Bridge\Doctrine\Transport\DoctrineTransport {"class":"Contao\\CoreBundle\\Messenger\\Message\\SearchIndexMessage","alias":"contao_prio_low","sender":"Symfony\\Component\\Messenger\\Bridge\\Doctrine\\Transport\\DoctrineTransport"} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"GET"}
[2024-08-01T22:25:00.169879+02:00] messenger.INFO: Stopping worker. {"transport_names":["contao_prio_high"]} {"request_uri":"https://abc.tld/contao?act=edit&do=themes&id=5&ref=9o84ouAE&rt=60b78e103353d3e.TvItpTqcZ4E_v6GSORdR2FPjm0OnKVQK0HCBt16WgvI.OqVpl2n5VexV0cjeC0UliiePozDlWxZduTrE7gzXurgEkVvOe6kM83ePzg&table=tl_module","request_method":"GET"}
```

If you then select the NEXT module type "Custom template" the "Internal Server Error" screen appears with this error message:

`Tried to list template options for the modern fragment type "frontend_module/root_page_dependent_modules" but could not find any template. Did you forget to create the default template?`

Didn't test on 5.3.11.

### PR DESCRIPTION
**Pull request by Toflar**
See https://github.com/contao/contao/issues/7404#issuecomment-2265143350

This is just a draft to outline the idea. Not sure we can merge this as-is, thus it's a draft :)

### CODE REVIEW

**Comment by leofeyer** [comment]
The changes look good to me, although I had hoped that we would only add one new method. Do you really think there is a use case for a public `getSubmittedRecord()` method?

**Comment by Toflar** [comment]
We have 3 use cases = 3 methods. Maybe someone wants to access only the data submitted in this very request so why not giving them the option?

**Review by fritzmg — DISMISSED** [review]
Draft seems logical to me 👍 
Though I think @ausi should also weigh in on this :)

**Comment by leofeyer** [comment]
> Maybe someone wants to access only the data submitted in this very request so why not giving them the option?

I just don‘t like adding new public API methods in a bugfix release unless they are necessary to fix the bug. Hence my question if we really want to make the `getSubmittedRecord()` method public.

**Comment by Toflar** [comment]
Imho, we should, yes.

**Comment by ausi** [comment]
I’m not sure about this. As this just reads the values from the POST data, it ignores all encoding settings from the widgets. And it also skips all validation checks. This was not the case with the active record AFAIK.

**Comment by Toflar** [comment]
Yes but that's exactly what was wrong. The old `activeRecord` looped over the widgets and - only in some cases - set the result of the widget to the object. So when calling it e.g. in a callback of the 4th field, it contained different information than when calling in a callback of the 5th field. It was non-deterministic and a mess. Hence we deprecated it.

**Comment by Toflar** [comment]
If you want to have the prepared submitted values, you can with our new `onbeforesubmit_callback` which uses the values after **all the widgets** have been processed. But if you are a callback yourself (like the `TemplateOptionsListener`), there is no deterministic way to get processed widget values. You only have 2 deterministic values: The ones in the POST request and the ones in the database.

**Comment by fritzmg** [comment]
Yes, even in Contao 4.13 we use `Input::post(…) ?: $dc->activeRecord->…`:

https://github.com/contao/contao/blob/56d1f5414ae0e1482ccdd2319490a8479338279b/core-bundle/src/Resources/contao/dca/tl_form.php#L495

**Comment by ausi** [comment]
In that case I think `getActiveRecord()` is misleading as it merges database values with POST values even though they are in a completely different format (potentially). This is a fact that I think we should not hide when this is used.

I’d prefer to remove the `getActiveRecord()` method and only add `getSubmittedRecord()`.

The simple usage would then look like `$dc->getSubmittedRecord()['type'] ?? $dc->getCurrentRecord()['type'] ?? null`. This would make it way more obvious what is happening and additionally adds the possibility to sanitize/encode the submitted data.

**Comment by Toflar** [comment]
That would be fine by me.

**Comment by fritzmg** [comment]
But that is still way less convenient than `$dc->activeRecord->…` - which we still use a lot in the core. Do we really want to rewrite all those instances to 

```php
$dc->getSubmittedRecord()[…] ?? $dc->getCurrentRecord()[…] ?? null
```

? I mean look at 

https://github.com/contao/contao/blob/4a6cac3f017f28760d45f1926bf8c220546c5d67/calendar-bundle/contao/dca/tl_calendar_events.php#L711-L776

for example. Without `$dc->activeRecord` or `$dc->getActiveRecord()` this will not be pretty.

// although in that particular case `$dc->getCurrentRecord()` should be fine, since it's an `onsubmit_callback` - but still.

**Comment by ausi** [comment]
> // although in that particular case `$dc->getCurrentRecord()` should be fine, since it's an `onsubmit_callback` - but still.

That’s exactly my point. If in any way possible, you should not rely on the POST data. In 99% of cases you should use `getCurrentRecord()`. All usages of `->activeRecord` should be checked if they can be replaced with `getCurrentRecord()` and only if not the `getSubmittedRecord()` fallback can be added. If we now add a `getActiveRecord()` method many people will just replace their calls to `->activeRecord` with it, which is wrong (and a potentially security risk) most of the time.

> But that is still way less convenient than `$dc->activeRecord->…` -

On purpose.

**Comment by fritzmg** [comment]
> In 99% of cases you should use `getCurrentRecord()`

Not sure that's true, I would expect most cases to have to use the "active" record, not the record stored in the database - as this has not been updated yet to the current state of the edit view.

**Comment by ausi** [comment]
> I would expect most cases to have to use the "active" record

But there is no such thing. Merging the submitted data with the database does not give us an “active” record. This would be misleading I think. Your example would not work with `getActiveRecord()`, because the date fields in the POST data are not converted into timestamps yet.

**Comment by fritzmg** [comment]
We deprecated `$dc->activeRecord` - but haven't actually removed all the usages yet (I guess because most of it is in legacy code). Changing everything (wherever applicable) to use the data from the `POST` request is a step backwards, as `$dc->activeRecord->type` would have had the properly processed value - albeit by chance (as the `type` field is processed before most other fields). Thus I think we need another solution - or un-deprecate `$dc->activeRecord` again, because I am not sure how feasible it is to create a solution where you could access the processed values of _all_ fields in a callback of a _single_ field (chicken and egg issue).

**Comment by aschempp** [comment]
I agree with @fritzmg, this will cause a lot of headaches. In my optioning (and hoping for it), all usages of `activeRecord` should be rewritten to use the `onbeforesubmit_callback`. For example with the alias, the save_callback should only **validate** the data. If the data is empty, it shouldn't create it. the `onbeforesubmit` callback should do that. But it will be a lot of work to rewrite that stuff…

**Comment by fritzmg** [comment]
But they cannot be rewritten to an `onbeforesubmit_callback`. Most of these need to be loaded when the edit mask is loaded, so either an `onload_callback` or field callback. And there you'll need `$dc->activeRecord` again.

**Comment by fritzmg** [comment]
> It was non-deterministic

It is still deterministic as the fields are processed in the order of the DCA configuration, i.e. the order of fields in `$GLOBALS['TL_DCA']['…']['fields']`. So if a field's callback depends on the active and processed value of another field, you can be sure that the value of that field is correct in `$dc->activeRecord` if that field comes before the callback's field. Many of our core field callbacks rely on this - and where they cannot rely on this, they currently use `Input::post()` - like the `generateAlias` callback of `tl_form`, as the `alias` field comes before the `jumpTo` field in the DCA configuration.

> and a mess.

I agree, but I currently do not see another way for the way the Contao back end works.

**Comment by Toflar** [comment]
No, it was not only non-deterministic because of that but also because it was not automated. Only a `$this->objActiveRecord->{$this->strField} = $this->varValue` call in the `DC_*` would make sure the value is set. And in some callbacks that is the case, in others it is not. Also, it's inconsistent across the drivers which makes things even worse.

**Comment by fritzmg** [comment]
> And in some callbacks that is the case, in others it is not.

Which callbacks do you mean? Only the `load_callback` will be executed before `$this->objActiveRecord->{$this->strField} = $this->varValue`, which is expected. But if your field comes after `type` you can be sure in your callback that `$dc->activeRecord->type` has the correctly processed value, after all callbacks.

> Also, it's inconsistent across the drivers which makes things even worse.

Yes, that is a general issue though, not just regarding the `activeRecord`.

**Comment by Toflar** [comment]
Stopping the discussion here because it leads to absolutely nowhere. We have a call for this.

**Comment by ausi** [comment]
> Only a `$this->objActiveRecord->{$this->strField} = $this->varValue` call in the `DC_*` would make sure the value is set.

And this is already done for every field that is loaded here: https://github.com/contao/contao/blob/287b2b03faf4d4910859782eed19ede95ebbd21f/core-bundle/contao/drivers/DC_Table.php#L2142 and for every field that is saved here: https://github.com/contao/contao/blob/287b2b03faf4d4910859782eed19ede95ebbd21f/core-bundle/contao/drivers/DC_Table.php#L3348

**Comment by fritzmg** [comment]
> We have a call for this.

The next call is in 3 weeks, imho we should make sure to fix #7404 before that. And currently I think the correct fix is to use `$dc->activeRecord->type` again - which implies that we need to un-deprecate `$dc->activeRecord`.

**Comment by aschempp** [comment]
I wanted to write down what I found out about the original issue. `DC_Table::getPalette()` actually uses the current POST data to find the active palette, without any validation. That's why it loads fields (the newly selected module) without updating the database. So anything that uses the same POST data would be equally unvalidated, but equally "how DC works" 😅 

> > It was non-deterministic
> 
> It is still deterministic as the fields are processed in the order of the DCA configuration, i.e. the order of fields in `$GLOBALS['TL_DCA']['…']['fields']`. 

I don't think that's true. It is the order of the palette, not the fields in the DCA. Which means the order varies depending on palette. If someone would move the type field to the end, the callback would fail.


> > Also, it's inconsistent across the drivers which makes things even worse.
> 
> Yes, that is a general issue though, not just regarding the `activeRecord`.

Not really an issue, because your DCA should know its driver and its limits.

**Comment by fritzmg** [comment]
> I don't think that's true. It is the order of the palette, not the fields in the DCA. Which means the order varies depending on palette. If someone would move the type field to the end, the callback would fail.

True, my mistake 👍

File: core-bundle/contao/drivers/DC_Table.php [hunk 1]
```diff
@@ -246,6 +246,31 @@ public function __construct($strTable, $arrModule=array())
...
+	 * Returns the current user changes merged with the database information.
+	 *
+	 * @return array<string, mixed>|null
+	 * @throws AccessDeniedException     if the current user has no read permission on the current record
+	 */
+	public function getActiveRecord(): array|null
+	{
+		$currentRecord = $this->getCurrentRecord();
+
+		if (null === $currentRecord)
```
Comment (leofeyer) [inline core-bundle/contao/drivers/DC_Table.php:259]:
```suggestion
		if (null === $currentRecord || [] === $this->arrSubmit)
```

**Review by ausi — DISMISSED** [review]

**Review by ausi — DISMISSED** [review]

**Review by fritzmg — DISMISSED** [review]

**Review by ausi — DISMISSED** [review]

File: core-bundle/contao/drivers/DC_Table.php [hunk 2]
```diff
@@ -246,6 +246,31 @@ public function __construct($strTable, $arrModule=array())
...
+		{
+			return null;
+		}
+
+		return array_merge($currentRecord, array_map(static function ($value) {
+			if (\is_string($value) || \is_bool($value) || \is_float($value) || \is_int($value) || $value === null)
```
Comment (aschempp) [inline core-bundle/contao/drivers/DC_Table.php]: I would add a comment to the database statement as where this code is originating from 🙈

**Review by aschempp — DISMISSED** [review]
There's probably a lot more places where we could use `getActiveRecord()` now, but I'm fine with not including this here 🙃

File: core-bundle/contao/drivers/DC_Table.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/contao/drivers/DC_Table.php]: Added in efe1b875dce296708c6b266e012cfaa60e9f0218.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**